### PR TITLE
Copy LruMap from scala-maxmind-iplookups

### PR DESCRIPTION
**Pull request by three**
This also adds the scalafmt and scoverage plugins to sbt  and adds a README.

### CODE REVIEW

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.maxmind.iplookups
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: wrong package

File: project/build.properties [hunk 2]
```diff
@@ -0,0 +1 @@
+sbt.version=1.1.4
```
Comment (BenFradet) [inline project/build.properties]: 1.1.6

File: build.sbt [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+lazy val root = project
+  .in(file("."))
+  .settings(
+    organization := "com.snowplowanalytics",
+    name := "lru-map",
+    version := "0.1.0",
+    description := "Simple LRU Map for caching",
+    scalaVersion := "2.11.12",
```
Comment (BenFradet) [inline build.sbt:20]: let's cross-compile with 2.11 and 2.12

File: .travis.yml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+language: scala
+scala:
+- 2.11.12
+- 2.12.5
```
Comment (BenFradet) [inline .travis.yml]: 2.12.6

File: README.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+          m2 <- calc(n-2)
+          _  <- lru.put(n, m1+m2)
+        } yield m1+m2
+      })
+  }
+}
+
+val result = (for {
+  lru <- LruMap.create[IO, BigInt, BigInt](500)
+  fib <- IO { new Fibonacci(lru) }
```
Comment (BenFradet) [inline README.md]: no need to wrap this in IO?

File: README.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+All impure methods and constructors are wrapped in [`cats.effect.Sync`][cats-sync].
+When you want to use the result in an impure environment you can use
+`unsafeRunSync` as shown above.
+
+## Note on Thread Safety
+
+Calls to `lruMap.get` and `LruMap.set` are not thread safe. Instead calls to
+lruMap should be synchronized by the user, or the same map should not be shared
+between threads.
```
Comment (BenFradet) [inline README.md]: I would replace this paragraph by concurrency concerns are left to the choice of `Sync`.

File: README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+          m1 <- calc(n-1)
+          m2 <- calc(n-2)
+          _  <- lru.put(n, m1+m2)
+        } yield m1+m2
+      })
+  }
+}
+
+val result = (for {
+  lru <- LruMap.create[IO, BigInt, BigInt](500)
```
Comment (BenFradet) [inline README.md]: we should mention what this 500 means

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    })
+
+  def put[F[_]: Sync, K, V](lruMap: LruMap[F, K, V])(k: K, v: V): F[Unit] =
+    lruMap.put(k, v)
+
+  def get[F[_]: Sync, K, V](lruMap: LruMap[F, K, V])(k: K): F[Option[V]] =
+    lruMap.get(k)
+
+  // initial capacity and load factor are the normal defaults for LinkedHashMap
+  def makeUnderlying[F[_]: Sync, K, V](maxSize: Int): F[ImpureLruMap[K, V]] = Sync[F].delay {
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: this should be private

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+// Based on com.twitter.util.LruMap
+// https://github.com/twitter/util/blob/develop/util-collection/src/main/scala/com/twitter/util/LruMap.scala
+
+object LruMap {
+  def create[F[_]: Sync, K, V](size: Int): F[LruMap[F, K, V]] =
+    Sync[F].flatMap(makeUnderlying[F, K, V](size))(underlying =>
+      Sync[F].delay {
+        new LruMap[F, K, V](size, underlying)
+    })
```
Comment (BenFradet) [inline src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala:27]: the new can be considered pure, no?

File: src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala (hunk 8 above)
Comment (BenFradet) [reply src/main/scala/com.snowplowanalytics.maxmind.lrumap/LruMap.scala]: and can be considered pure, no?

File: README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
+# Scala LruMap
```
Comment (BenFradet) [inline README.md:1]: you'll need to add the necessary badges

**Comment by three** [comment]
@BenFradet I don't think constructors can be considered pure since they return a different reference each time they're called. In Haskell the result of creating a mutable reference like with `newSTRef` or `newArray` would are wrapped in a monad since it's not referentially transparent. Although I'm not sure if things are differently in Scala?

**Comment by three** [comment]
Should I also copy over the deploy code from iplookups?

**Comment by BenFradet** [comment]
>@BenFradet I don't think constructors can be considered pure since they return a different reference each time they're called. In Haskell the result of creating a mutable reference like with newSTRef or newArray would are wrapped in a monad since it's not referentially transparent. Although I'm not sure if things are differently in Scala?

Yes but here, in the examples I pointed out, you're not creating a mutable reference, you're just performing memory allocation which we don't consider to be a side-effect.

>Should I also copy over the deploy code from iplookups?

yup

**Pull request closed without merging**